### PR TITLE
Batch saving: push to redis queue in post

### PR DESCRIPTION
**Pull request by sastels**
# Summary | Résumé

Add a feature flag and code for pushing to the redis queue in the POST

# Test instructions | Instructions pour tester la modification

This is basically a.rough in - not hooked up to the real queue and hidden behind a feature flag

# Release Instructions | Instructions pour le déploiement

none (code path disabled for now)

# Reviewer checklist | Liste de vérification du réviseur

This is a suggested checklist of questions reviewers might ask during their
review | Voici une suggestion de liste de vérification comprenant des questions
que les réviseurs pourraient poser pendant leur examen :

- [ ] Is the code maintainable? | Est-ce que le code peut être maintenu?
- [ ] Have you tested it? | L’avez-vous testé?
- [ ] Are there automated tests? | Y a-t-il des tests automatisés?
- [ ] Does this cause automated test coverage to drop? | Est-ce que ça entraîne
      une baisse de la quantité de code couvert par les tests automatisés?
- [ ] Does this break existing functionality? | Est-ce que ça brise une
      fonctionnalité existante?
- [ ] Does this change the privacy policy? | Est-ce que ça entraîne une
      modification de la politique de confidentialité?
- [ ] Does this introduce any security concerns? | Est-ce que ça introduit des
      préoccupations liées à la sécurité?
- [ ] Does this significantly alter performance? | Est-ce que ça modifie de
      façon importante la performance?
- [ ] What is the risk level of using added dependencies? | Quel est le degré de
      risque d’utiliser des dépendances ajoutées?
- [ ] Should any documentation be updated as a result of this? (i.e. README
      setup, etc.) | Faudra-t-il mettre à jour la documentation à la suite de ce
      changement (fichier README, etc.)?

### CODE REVIEW

File: tests/app/v2/notifications/test_post_notifications.py [hunk 1]
```diff
@@ -132,8 +132,51 @@ def test_post_sms_notification_with_persistance_in_celery_returns_201(
...
+            )
+            in resp_json["template"]["uri"]
+        )
+        assert not resp_json["scheduled_for"]
+        assert mocked_redis_publish.called
+
+
 def test_post_sms_notification_uses_inbound_number_as_sender(notify_api, client, notify_db_session, mocker):
     notify_api.config["FF_NOTIFICATION_CELERY_PERSISTENCE"] = False
+    notify_api.config["FF_REDIS_BATCH_SAVING"] = False
```
Comment (sastels) [inline tests/app/v2/notifications/test_post_notifications.py:179]: our tests are colliding with each other, in particular with changing config settings (ie these feature flags). We should get this sorted, with proper setup and teardown if necessary.

File: app/v2/notifications/post_notifications.py [hunk 2]
```diff
@@ -84,6 +84,13 @@
 )
 
 
+# TODO: replace with the real thing
+class RedisQueue:
```
Comment (jimleroyer) [inline app/v2/notifications/post_notifications.py:88]: We opened that PR for you to work easily with an abstraction:
https://github.com/cds-snc/notification-api/pull/1437

Please review so we can work in tandem.

File: app/v2/notifications/post_notifications.py [hunk 3]
```diff
@@ -84,6 +84,13 @@
...
+    @staticmethod
+    def publish(notification):
```
Comment (jimleroyer) [inline app/v2/notifications/post_notifications.py:90]: The latest signature is:
```
def publish(self, serializable: Serializable):
```

The latest is there:
https://github.com/cds-snc/notification-api/blob/d97138603a66ce251185516c53b5a59b4ffc196a/app/queue.py#L212

File: app/v2/notifications/post_notifications.py [hunk 4]
```diff
@@ -291,6 +298,10 @@ def process_sms_or_email_notification(*, form, notification_type, api_key, templ
         )
         persist_scheduled_notification(notification.id, form["scheduled_for"])
 
+    elif current_app.config["FF_REDIS_BATCH_SAVING"] and not simulated:
+        RedisQueue.publish(notification)
+        current_app.logger.info(f"{notification_type} {notification['id']} sent to RedisQueue")
```
Comment (jimleroyer) [inline app/v2/notifications/post_notifications.py:303]: We should add a task to revert our logger messages to debug level eventually...

File: tests/app/v2/notifications/test_post_notifications.py (hunk 1 above)
Comment (jimleroyer) [reply tests/app/v2/notifications/test_post_notifications.py:179]: Yeah, maintenance will get out of control as we would have tests that should be multiplied for each logical branch derived by these env vars tweak.

**Review by jimleroyer — APPROVED** [review]
LGTM!

(please merge our branch so we can work with "almost" the same queue implementation in your future changes! 🙏 )

**Comment by jzbahrai** [comment]
@sastels can we change the base of this branch?

File: app/config.py [hunk 5]
```diff
@@ -426,6 +426,7 @@ class Config(object):
     # feature flag to toggle persistance of notification in celery instead of the API
     FF_NOTIFICATION_CELERY_PERSISTENCE = os.getenv("FF_NOTIFICATION_CELERY_PERSISTENCE", False)
     FF_BATCH_INSERTION = os.getenv("FF_BATCH_INSERTION", False)
+    FF_REDIS_BATCH_SAVING = os.getenv("FF_REDIS_BATCH_SAVING", False)
```
Comment (jzbahrai) [inline app/config.py:429]: is there a manifest PR too?

File: app/v2/notifications/post_notifications.py [hunk 6]
```diff
@@ -291,6 +298,14 @@ def process_sms_or_email_notification(*, form, notification_type, api_key, templ
         )
         persist_scheduled_notification(notification.id, form["scheduled_for"])
 
+    elif current_app.config["FF_REDIS_BATCH_SAVING"] and not simulated:
+        redis_queue = RedisQueue()
+        if notification_type == SMS_TYPE:
+            redis_queue.publish(encrypted_notification_data)
```
Comment (jzbahrai) [inline app/v2/notifications/post_notifications.py:304]: there will be two separate redis queues here?

**Review by jzbahrai — APPROVED** [review]

File: app/config.py (hunk 5 above)
Comment (sastels) [reply app/config.py:429]: no, not yet! let me go do that...

File: app/v2/notifications/post_notifications.py (hunk 6 above)
Comment (sastels) [reply app/v2/notifications/post_notifications.py:304]: yeah, exactly.

File: app/config.py (hunk 5 above)
Comment (sastels) [reply app/config.py:429]: https://github.com/cds-snc/notification-manifests/pull/863

**Pull request merged**